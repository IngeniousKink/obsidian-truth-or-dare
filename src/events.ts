export type GameEvent = 
    | CreateActorEvent
    | DeleteActorEvent
    | ChangeActorNameEvent
    | ChangeActorInventoryAddEvent
    | ChangeActorInventoryRemoveEvent
    | DrawCardEvent
    | SkipCardEvent
    | CompleteCardEvent
    | AssignSlotEvent
    | UnassignSlotEvent
    | AdvanceStackEvent;

export type WithTimestamp = {
    timestamp: number;
};

export type TimestampedEvent = GameEvent & WithTimestamp;

export type CreateActorEvent = {
    type: 'create_actor';
    actorId: string;
};

export type DeleteActorEvent = {
    type: 'delete_actor';
    actorId: string;
};

export type ChangeActorNameEvent = {
    type: 'change_actor_name';
    actorId: string;
    value: string;
};

export type ChangeActorInventoryAddEvent = {
    type: 'change_actor_inventory_add';
    actorId: string;
    value: string;
};

export type ChangeActorInventoryRemoveEvent = {
    type: 'change_actor_inventory_remove';
    actorId: string;
    value: string;
};

export type DrawCardEvent = {
    type: 'draw_card';
    cardRef: string;
};

export type SkipCardEvent = {
    type: 'skip_card';
    cardRef: string;
};

export type CompleteCardEvent = {
    type: 'complete_card';
    cardRef: string;
};

export type AssignSlotEvent = {
    type: 'assign_slot';
    actorId: string;
    slotIndex: number;
};

export type UnassignSlotEvent = {
    type: 'unassign_slot';
    slotIndex: number;
};

export type AdvanceStackEvent = {
    type: 'advance_stack';
};

export function timestampEvent<T extends GameEvent>(event: T): T & WithTimestamp {
    return {
        ...event,
        timestamp: Date.now()
    };
}

export function createActor(): CreateActorEvent {
    const actorId = (
        (Date.now() - 1700000000000) /** for good measure */
        .toString(32)
        .split('')
        .reverse()
        .join('')
    );

    return {
        type: 'create_actor',
        actorId
    };
}

export function deleteActor(actorId: string): DeleteActorEvent {
    return {
        type: 'delete_actor',
        actorId
    };
}

export function changeActorName(actorId: string, value: string): ChangeActorNameEvent {
    return {
        type: 'change_actor_name',
        actorId,
        value
    };
}

export function changeActorInventoryAdd(actorId: string, inventoryToAdd: string): ChangeActorInventoryAddEvent {
    return {
        type: 'change_actor_inventory_add',
        actorId,
        value: inventoryToAdd
    };
}

export function changeActorInventoryRemove(actorId: string, inventoryToRemove: string): ChangeActorInventoryRemoveEvent {
    return {
        type: 'change_actor_inventory_remove',
        actorId,
        value: inventoryToRemove
    };
}

export function assignSlot(actorId: string, slotIndex: number): AssignSlotEvent {
    return {
        type: 'assign_slot',
        actorId,
        slotIndex
    };
}

export function unassignSlot(slotIndex: number): UnassignSlotEvent {
    return {
        type: 'unassign_slot',
        slotIndex
    };
}

export function drawCard(cardRef: string): DrawCardEvent {
    return {
        type: 'draw_card',
        cardRef,
    };
}

export function skipCard(cardRef: string): SkipCardEvent {
    return {
        type: 'skip_card',
        cardRef
    };
}

export function completeCard(cardRef: string): CompleteCardEvent {
    return {
        type: 'complete_card',
        cardRef
    };
}

export function advanceStack(): AdvanceStackEvent {
    return {
        type: 'advance_stack',
    };
}

export const serializeEventToCodeBlock = (event: TimestampedEvent) => {
    let eventString = `\`\`\`truth-or-dare:event\n`;

    // Ensure type and timestamp are always at the top
    eventString += `type:${event.type}\n`;
    eventString += `timestamp:${event.timestamp}\n`;

    const eventWithStrKeys: { [key: string]: any } = event;

    for (const key in eventWithStrKeys) {
        if (key !== "type" && key !== "timestamp") {
            eventString += `${key}:${eventWithStrKeys[key]}\n`;
        }
    }

    eventString += `\`\`\`\n`;

    return eventString;
};
