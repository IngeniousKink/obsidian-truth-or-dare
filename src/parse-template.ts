import type { Root, Html, Heading, List, ListItem, Paragraph, Text, Image } from "mdast";
import { parseCard } from "./parse-card.js";
import type { ParsedCard } from "./parse-card.js";
import { PhrasingContent } from "node_modules/mdast-util-from-markdown/lib/index.js";
import { parseEmbeds } from "./parse-embeds.js";

export type CardWithRef = ParsedCard & { ref: string; category?: string };

// Define a Stack type with a name, reference and an array of cards
export type Stack = { name: string; ref: string; cards: CardWithRef[]; };

// Define a GameTemplate type with an array of stacks
export type GameTemplate = { stacks: Stack[]; };
// ChildNode type can be either a Heading or a List
type ChildNode = Heading | List;

/**
 * This function takes a markdown abstract syntax tree (AST) and converts it into a game template.
 * The game template is represented as a GameTemplate, where each stack is an object with a name, reference and an array of cards.
 * Each heading in the markdown creates a new stack in the GameTemplate, and each list item under that heading
 * creates a new card in that stack.
 * 
 * For example, given the following markdown:
 * 
 * # Heading 1
 * - Item 1
 * - Item 2
 * 
 * # Heading 2
 * - Item 3
 * 
 * The resulting GameTemplate would be:
 * 
 * {
 *   stacks: [
 *     {
 *       name: "Heading 1",
 *       ref: "#Heading 1",
 *       cards: [
 *         { ref: "#Heading 1^0", text: "Item 1" },
 *         { ref: "#Heading 1^1", text: "Item 2" }
 *       ]
 *     },
 *     {
 *       name: "Heading 2",
 *       ref: "#Heading 2",
 *       cards: [
 *         { ref: "#Heading 2^0", text: "Item 3" }
 *       ]
 *     }
 *   ]
 * }
 */
export function convertMarkdownToGameTemplate(node: Root): GameTemplate {
  return node.children.reduce<GameTemplate>(
    (gameTemplate, child) => processMarkdownNode(gameTemplate, child as ChildNode),
    { stacks: [] }
  );
}

// Process a markdown node, which can either be a heading or a list
function processMarkdownNode(gameTemplate: GameTemplate, child: ChildNode): GameTemplate {
  if (child.type === 'heading') {
    const headingText = extractHeadingText(child);
    gameTemplate.stacks.push({ name: headingText, ref: "#" + headingText, cards: [] });
    return gameTemplate;
  }

  if (child.type === 'list') {
    let stack = getLastStack(gameTemplate);
    if (!stack) {
      stack = {
        name: "",
        ref: "",
        cards: extractCardsFromList(child, ""),
      };
      gameTemplate.stacks.push(stack);
    }
    stack.cards = extractCardsFromList(child, stack.ref);
    return gameTemplate;
  }

  return gameTemplate;
}

// Extract the text from a heading node
function extractHeadingText(node: Heading): string {
  const textNode = node.children[0] as Text;
  return textNode.value || '';
}

// Get the last stack from the game template
function getLastStack(gameTemplate: GameTemplate): Stack {
  return gameTemplate.stacks[gameTemplate.stacks.length - 1];
}

// Extract cards from a list node
function extractCardsFromList(listNode: List, stackRef: string): CardWithRef[] {
  let refCounter = 0;
  return listNode.children.reduce<CardWithRef[]>((cards, listItem: ListItem) => {
    const listItemCards = extractCardsFromListItem(listItem, refCounter++, stackRef);
    return [...cards, ...listItemCards];
  }, []);
}

// Extract cards from a list item
function extractCardsFromListItem(listItem: ListItem, refCounter: number, stackRef: string): CardWithRef[] {
  return listItem.children.reduce<CardWithRef[]>((cards, paragraph: Paragraph | Html) => {
    const paragraphCard = extractCardFromParagraph(paragraph, refCounter, stackRef);
    return [...cards, paragraphCard];
  }, []);
}

export function extractCardFromParagraph(paragraph: Paragraph | Html, refCounter: number, stackRef: string): CardWithRef {
  let children : PhrasingContent[]  = [];

  if (paragraph.type == 'html') {
    children = [paragraph];
  } else {
    children = paragraph.children;
  }

  return {
    ...parseCard(parseEmbeds(children)),
    ref: stackRef + '^' + refCounter,
  };
}