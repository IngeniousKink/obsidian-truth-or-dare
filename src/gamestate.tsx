import { CreateActorEvent, DrawCardEvent, GameEvent, TimestampedEvent, DeleteActorEvent, ChangeActorNameEvent, AssignSlotEvent, UnassignSlotEvent, CompleteCardEvent } from "./events.js";
import { ParsedCard } from "./parse/parse-card.js";
import { CardWithRef, GameTemplate } from "./parse/parse-template.js";
import seedrandom from 'seedrandom';

export type Actor = {
    id: string;
    name: string;
    inventory: string[]
};

export type GameState = {
    template: GameTemplate;

    events: TimestampedEvent[];
    seed: string;

    previousCards: string[];

    actors: Actor[];

    displayedCard?: string;
    allocation: (string | null)[];

    preview: {
        displayedCard?: string,
        allocation: (string | null)[],
    }
};

export function createGameState(gameTemplate: GameTemplate, gameEvents: TimestampedEvent[], seed?: string): GameState {
    let gameState: GameState = {
        template: gameTemplate,
        events: gameEvents,
        previousCards: [],
        actors: [],
        allocation: [],
        preview: {
            allocation: []
        },
        seed: seed || JSON.stringify(gameTemplate) + JSON.stringify(gameEvents)
    };

    for (const event of gameEvents) {
        gameState = applyEventToGameState(gameState, event);
    }

    return gameState;
}


export function findCardInGameTemplate(gameTemplate: GameTemplate, searchRef: string): CardWithRef | null {
    for (const stack of gameTemplate.stacks) {
        for (const card of stack.cards) {
            if (card.ref == searchRef) {
                return card;
            }
        }
    }
    return null;
}

export function getAllCards(gameTemplate: GameTemplate): CardWithRef[] {
    return gameTemplate.stacks.reduce((arr, stack) => arr.concat(stack.cards), [] as CardWithRef[]);
}

export function selectRandomCard(cards: CardWithRef[], gameState: GameState): CardWithRef | null {
    if (cards.length === 0) return null;

    const rng = seedrandom(gameState.seed);
    const randomIndex = Math.floor(rng() * cards.length);
    return cards[randomIndex];
}

export function selectRandomAvailableCard(gameState: GameState): CardWithRef | null {
    const allCards = getAvailableCards(gameState);
    return selectRandomCard(allCards, gameState);
}

export function getAvailableCards(gameState: GameState): CardWithRef[] {
    const allCards = getAllCards(gameState.template);
    const unavailableCards = [gameState.displayedCard, ...gameState.previousCards];
    return allCards.filter(card => !unavailableCards.includes(card.ref));
}

export function getCategories(card: ParsedCard): string[] {
    return card.annotations
        .map(annotation => annotation?.category)
        .filter(category => category !== undefined) as string[];
}

export function selectCategories(gameState: GameState): string[] {
    const allCards = getAllCards(gameState.template);

    const categories = allCards.flatMap(getCategories)

    return [...new Set(categories)];
}

export function selectCardsByCategory(gameState: GameState): { [key: string]: CardWithRef[] } {
    const categories = selectCategories(gameState);
    const availableCards = getAvailableCards(gameState);
    const cardsByCategory: { [key: string]: CardWithRef[] } = {};

    for (const category of categories) {
        cardsByCategory[category] = availableCards
            .filter(card => getCategories(card).includes(category))
    }

    return cardsByCategory;
}

export function selectCardByRef(gameState: GameState, ref: string | undefined): CardWithRef | null {
    if (!ref) return null;

    const allCards = getAllCards(gameState.template);
    for (const card of allCards) {
        if (card.ref === ref) {
            return card;
        }
    }
    return null;
}

function getNextActorId(gameState: GameState, lastActorId: string): string | null {
    let currentActorIndex = gameState.actors.findIndex(
        actor => actor.id === lastActorId);

    if (currentActorIndex === -1) {
        return null;
    }

    if (gameState.actors.length < 1) {
        return null;
    } 

    const nextActorIndex = (currentActorIndex + 1) % gameState.actors.length;
    return gameState.actors[nextActorIndex].id;
}

function maintainAllocation(gameState: GameState): GameState {
    const lastActorId = gameState.preview.allocation[0] || gameState.actors[0]?.id;

    if (!lastActorId) {
        return {
            ...gameState,
            allocation: [],
            preview: {
                ...gameState.preview,
                allocation: [],
            },
        };
    }

    const nextPlayerId = getNextActorId(gameState, lastActorId);
    const newAllocation = [...gameState.allocation];

    if (lastActorId) {
        newAllocation[0] = lastActorId;
    }

    let previewAllocation: string[] = [];
    if (newAllocation.length > 0 && nextPlayerId) {
        previewAllocation = [nextPlayerId];
    }

    return {
        ...gameState,
        allocation: newAllocation,
        preview: {
            ...gameState.preview,
            allocation: previewAllocation,
        },
    };
}

function handleDrawCardEvent(gameState: GameState, event: DrawCardEvent): GameState {
    const card = findCardInGameTemplate(gameState.template, event.cardRef);
    if (!card) return gameState;

    let previousCards = [...gameState.previousCards];
    if (gameState.displayedCard) {
        previousCards.push(gameState.displayedCard);
    }

    let updatedGameState = maintainAllocation(gameState);

    return {
        ...updatedGameState,
        previousCards: previousCards,
        displayedCard: card.ref,
    };
}



function handleCompleteCardEvent(gameState: GameState, event: CompleteCardEvent): GameState {
    return {
        ...gameState,
        allocation: [],
        displayedCard: undefined,
    };
}


function handleCreateActor(state: GameState, event: CreateActorEvent): GameState {
    return {
        ...state,
        actors: [
            ...state.actors,
            {
                id: event.actorId,
                name: `Player ${event.actorId}`,
                inventory: []
            }
        ]
    };
}

function handleDeleteActor(state: GameState, event: DeleteActorEvent): GameState {
    let newState = { ...state };

    newState.allocation.forEach((actorId, slotIndex) => {
        if (actorId === event.actorId) {
            newState = handleUnassignSlot(newState, { type: 'unassign_slot', slotIndex });
        }
    });

    return {
        ...newState,
        actors: newState.actors.filter(actor => actor.id !== event.actorId),
    };
}

function handleChangeActorName(state: GameState, event: ChangeActorNameEvent): GameState {
    return {
        ...state,
        actors: state.actors.map(actor =>
            actor.id === event.actorId
                ? { ...actor, name: event.value }
                : actor
        )
    };
}

function handleAssignSlot(state: GameState, event: AssignSlotEvent): GameState {
    const newAllocation = [...state.allocation];

    // Fill the array with null up to the slotIndex
    while (newAllocation.length <= event.slotIndex) {
        newAllocation.push(null);
    }

    newAllocation[event.slotIndex] = event.actorId;

    return {
        ...state,
        allocation: newAllocation,
    };
}

function handleUnassignSlot(state: GameState, event: UnassignSlotEvent): GameState {
    const newAllocation = [...state.allocation];
    newAllocation[event.slotIndex] = null;

    return {
        ...state,
        allocation: newAllocation,
    };
}


function applyEventToGameState(gameState: GameState, event: GameEvent): GameState {
    switch (event.type) {
        case 'draw_card':
            return handleDrawCardEvent(gameState, event as DrawCardEvent);
        case 'complete_card':
            return handleCompleteCardEvent(gameState, event as CompleteCardEvent);
        case 'create_actor':
            return handleCreateActor(gameState, event as CreateActorEvent);
        case 'delete_actor':
            return handleDeleteActor(gameState, event as DeleteActorEvent);
        case 'change_actor_name':
            return handleChangeActorName(gameState, event as ChangeActorNameEvent);
        case 'assign_slot':
            return handleAssignSlot(gameState, event as AssignSlotEvent);
        case 'unassign_slot':
            return handleUnassignSlot(gameState, event as UnassignSlotEvent);
        default:
            return gameState;
    }
}
