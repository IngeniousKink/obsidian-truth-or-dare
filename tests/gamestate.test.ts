import { TimestampedEvent } from '@obsidian-truth-or-dare/events.js';
import { createGameState, findCardInGameTemplate, getAllCards, getAvailableCards, selectCardsByCategory, GameState } from '../src/gamestate.js';
import { CardWithRef, GameTemplate, Stack } from '@obsidian-truth-or-dare/parse/parse-template.js';
import { beforeEach, describe, expect, test } from 'vitest';

describe('Game State', () => {
  const mockGameTemplate = {
    stacks: [
      {
        name: 'stack1',
        ref: 'stackRef',
        cards: [
          { ref: 'card1', text: 'card1Text', annotations: [] },
          { ref: 'card2', text: 'card2Text', annotations: [] },
          { ref: 'card3', text: 'card3Text', annotations: [] }
        ]
      }
    ]
  }

  const mockGameEvents: TimestampedEvent[] = [
    { type: 'draw_card', cardRef: 'card1', timestamp: 0},
    { type: 'complete_card', cardRef: 'card1', timestamp: 0},
    { type: 'draw_card', cardRef: 'card2', timestamp: 0},
  ]

  test('createGameState', () => {
    const result = createGameState(mockGameTemplate, mockGameEvents);
    expect(result).toMatchSnapshot();
  });

  test('findCardInGameTemplate', () => {
    const result = findCardInGameTemplate(mockGameTemplate, 'card1');
    expect(result).toMatchSnapshot();
  });

  test('getAllCards', () => {
    const result = getAllCards(mockGameTemplate);
    expect(result).toMatchSnapshot();
  });

  test('getAvailableCards', () => {
    const mockGameState = createGameState(mockGameTemplate, mockGameEvents);
    const result = getAvailableCards(mockGameState);
    expect(result).toMatchSnapshot();
  });
});


describe('selectCardsByCategory', () => {
  let gameState: GameState;
  let gameTemplate: GameTemplate;
  let cards: CardWithRef[];
  let stack: Stack;

  beforeEach(() => {
    cards = [
      { ref: '1', text: 'Card 1', annotations: [{category: 'Category 1'}] },
      { ref: '2', text: 'Card 2', annotations: [{category: 'Category 2'}] },
      { ref: '3', text: 'Card 3', annotations: [{category: 'Category 3'}] },
      { ref: '4', text: 'Card 4', annotations: [{category: 'Category 1'}] },
      { ref: '5', text: 'Card 5', annotations: [{category: 'Category 2'}] },
      { ref: '6', text: 'Card 6', annotations: [{category: 'Category 3'}] },
    ];
    stack = { name: 'Stack 1', ref: '#Stack 1', cards: cards };
    gameTemplate = { stacks: [stack] };
    gameState = createGameState(gameTemplate, []);
  });

  test('should return cards grouped by category', () => {
    const result = selectCardsByCategory(gameState);
    expect(result).toEqual({
      'Category 1': [
        { ref: '1', text: 'Card 1', annotations: [{category: 'Category 1'}] },
        { ref: '4', text: 'Card 4', annotations: [{category: 'Category 1'}] }
      ],
      'Category 2': [
        { ref: '2', text: 'Card 2', annotations: [{category: 'Category 2'}] },
        { ref: '5', text: 'Card 5', annotations: [{category: 'Category 2'}] }
      ],
      'Category 3': [
        { ref: '3', text: 'Card 3', annotations: [{category: 'Category 3'}] },
        { ref: '6', text: 'Card 6', annotations: [{category: 'Category 3'}] }
      ],
    });
  });

  test('should return an empty object if there are no categories', () => {
    const result = selectCardsByCategory({
      ...gameState,
      template: {
        ...gameTemplate,
        stacks: [],
      }
    });
    expect(result).toEqual({});
  });

  test('should return an empty array for each category if there are no available cards', () => {
    const result = selectCardsByCategory({
      ...gameState,
      previousCards: ['1', '2', '3', '4', '5', '6']
    });
    expect(result).toEqual({ 'Category 1': [], 'Category 2': [], 'Category 3': [] });
  });


  test('should only include cards that are available', () => {
    gameState = {
      ...gameState,
      previousCards: ['4']
    };
    const result = selectCardsByCategory(gameState);
    expect(result['Category 1']).not.toContain('4');
  });

  test('should not include categories that have no available cards', () => {
    gameState = {
      ...gameState,
      previousCards: ['1', '4']
    };
    const result = selectCardsByCategory(gameState);
    expect(result).not.toHaveProperty('Category1');
  });
});

