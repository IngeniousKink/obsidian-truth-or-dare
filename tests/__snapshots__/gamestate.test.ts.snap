// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game State createGameState 1`] = `
{
  "displayedCard": "card2",
  "events": [
    {
      "card": "card1",
      "timestamp": 0,
      "type": "card-draw",
    },
    {
      "card": "card2",
      "timestamp": 0,
      "type": "card-draw",
    },
  ],
  "previousCards": [
    "card1",
  ],
  "seed": "{"stacks":[{"name":"stack1","ref":"stackRef","cards":[{"ref":"card1","text":"card1Text","annotations":[]},{"ref":"card2","text":"card2Text","annotations":[]},{"ref":"card3","text":"card3Text","annotations":[]}]}]}[{"type":"card-draw","card":"card1","timestamp":0},{"type":"card-draw","card":"card2","timestamp":0}]",
  "template": {
    "stacks": [
      {
        "cards": [
          {
            "annotations": [],
            "ref": "card1",
            "text": "card1Text",
          },
          {
            "annotations": [],
            "ref": "card2",
            "text": "card2Text",
          },
          {
            "annotations": [],
            "ref": "card3",
            "text": "card3Text",
          },
        ],
        "name": "stack1",
        "ref": "stackRef",
      },
    ],
  },
}
`;

exports[`Game State findCardInGameTemplate 1`] = `
{
  "annotations": [],
  "ref": "card1",
  "text": "card1Text",
}
`;

exports[`Game State getAllCards 1`] = `
[
  {
    "annotations": [],
    "ref": "card1",
    "text": "card1Text",
  },
  {
    "annotations": [],
    "ref": "card2",
    "text": "card2Text",
  },
  {
    "annotations": [],
    "ref": "card3",
    "text": "card3Text",
  },
]
`;

exports[`Game State getAvailableCards 1`] = `
[
  {
    "annotations": [],
    "ref": "card3",
    "text": "card3Text",
  },
]
`;
